relayer/chains/wasm/provider.go

```diff
@@ -0,0 +1,746 @@
+		resultEventChan, err := httpClient.Subscribe(newCtx, "tx-result-waiter", query)
+		if err != nil {
+			txRes <- &types.TxResultChan{
+				TxResult: nil, Error: err,
+			}
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+			txRes <- &types.TxResultChan{
+				TxResult: nil, Error: ctx.Err(),
+			}
+			return
+		case e := <-resultEventChan:
+			eventDataJSON, err := json.Marshal(e.Data)
+			if err != nil {
+				txRes <- &types.TxResultChan{
+					TxResult: nil, Error: err,
+				}
+				return
+			}
+
+			txWaitRes := new(types.TxResultWaitResponse)
+			if err := json.Unmarshal(eventDataJSON, txWaitRes); err != nil {
+				txRes <- &types.TxResultChan{
+					TxResult: nil, Error: err,
+				}
+				return
+			}
+			if uint32(txWaitRes.Result.Code) != types.CodeTypeOK {
+				txRes <- &types.TxResultChan{
+					Error: fmt.Errorf(txWaitRes.Result.Log),
+					TxResult: &relayTypes.TxResponse{
+						Height:    txWaitRes.Height,
+						TxHash:    txHash,
+						Codespace: txWaitRes.Result.Codespace,
+						Code:      relayTypes.ResponseCode(txWaitRes.Result.Code),
+						Data:      string(txWaitRes.Result.Data),
+					},
+				}
+				return
+			}
+
+			txRes <- &types.TxResultChan{
+				TxResult: &relayTypes.TxResponse{
+					Height:    txWaitRes.Height,
+					TxHash:    txHash,
+					Codespace: txWaitRes.Result.Codespace,
+					Code:      relayTypes.ResponseCode(txWaitRes.Result.Code),
+					Data:      string(txWaitRes.Result.Data),
+				},
+			}
+		}
+	}(txResChan)
+	return txResChan
+}
+
+func (p *Provider) MessageReceived(ctx context.Context, key *relayTypes.MessageKey) (bool, error) {
+	queryMsg := &types.QueryReceiptMsg{
+		GetReceipt: &types.GetReceiptMsg{
+			SrcNetwork: key.Src,
+			ConnSn:     strconv.FormatUint(key.Sn, 10),
+		},
+	}
+	rawQueryMsg, err := json.Marshal(queryMsg)
+	if err != nil {
+		return false, err
+	}
+
+	res, err := p.client.QuerySmartContract(ctx, p.cfg.Contracts[relayTypes.ConnectionContract], rawQueryMsg)
+	if err != nil {
+		p.logger.Error("failed to check if message is received: ", zap.Error(err))
+		return false, err
+	}
+
+	receiptMsgRes := types.QueryReceiptMsgResponse{}
+	return receiptMsgRes.Status, json.Unmarshal(res.Data, &receiptMsgRes.Status)
+}
+
+func (p *Provider) QueryBalance(ctx context.Context, addr string) (*relayTypes.Coin, error) {
+	coin, err := p.client.GetBalance(ctx, addr, p.cfg.Denomination)
+	if err != nil {
+		p.logger.Error("failed to query balance: ", zap.Error(err))
+		return nil, err
+	}
+	return &relayTypes.Coin{
+		Denom:  coin.Denom,
+		Amount: coin.Amount.Uint64(),
+	}, nil
+}
+
+func (p *Provider) ShouldReceiveMessage(ctx context.Context, message *relayTypes.Message) (bool, error) {
+	return true, nil
+}
+
+func (p *Provider) ShouldSendMessage(ctx context.Context, message *relayTypes.Message) (bool, error) {
+	return true, nil
+}
+
+func (p *Provider) GenerateMessages(ctx context.Context, messageKey *relayTypes.MessageKeyWithMessageHeight) ([]*relayTypes.Message, error) {
```

Comment:
Finality processor is not needed in cosmos based chains.

Reply:
Needed to ensure packet are delivered irrespective of finality.